src/hotspot/share/code/nmethod.cpp

```diff
@@ -2711,15 +2730,15 @@ PcDesc* PcDescContainer::find_pc_desc_internal(address pc, bool approximate, con
 #undef assert_LU_OK
 
   if (match_desc(upper, pc_offset, approximate)) {
-    assert(upper == linear_search(search, pc_offset, approximate), "search ok");
+    assert(upper == linear_search(pc_offset, approximate, lower_incl, upper_incl), "search mismatch");
     if (!Thread::current_in_asgct()) {
       // we don't want to modify the cache if we're in ASGCT
       // which is typically called in a signal handler
       _pc_desc_cache.add_pc_desc(upper);
     }
```

Comment:
The special case for ASGCT, along with ThreadWXEnable changes here:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/sharedRuntime.cpp#L484
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/sharedRuntime.cpp#L1755
https://github.com/openjdk/jdk/blob/master/src/hotspot/os/posix/signals_posix.cpp#L617
look like they will no longer be needed.  I suggest filing a follow-up RFE for an ASGCT expert to take.

Reply:
Thank you, @dean-long, for pointing this. 

[JDK-8302736](https://bugs.openjdk.org/browse/JDK-8302736) added 2 ThreadWXEnable in sharedRuntime.cpp
And later [JDK-8316392](https://bugs.openjdk.org/browse/JDK-8316392) added it to `PcDescCache::add_pc_desc()`
What a mess ...

We can now safely remove (after testing) all ThreadWXEnable which guards calls to `add_pc_desc()`
And file separate RFE for runtime to look on `!Thread::current_in_asgct()`

Is this what you are suggesting? I will do it.

Reply:
ASGCT code was added by [JDK-8304725](https://github.com/openjdk/jdk/commit/d8af7a6014055295355a1242db6c2872299c6398) before [JDK-8316392](https://bugs.openjdk.org/browse/JDK-8316392) added ThreadWXEnable to PcDescCache::add_pc_desc().

As you suggested I will file follow up RFE to remove that code.

Reply:
RFE: https://bugs.openjdk.org/browse/JDK-8331012